lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,469 @@
+            vault_address: The address of the Vault server.
+            approle_id: The approle id to use when authenticating with the Vault server.
+            approle_secret_id: The approle secret id to use when authenticating with the Vault server.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
+            logger.warning(
+                "Attempting to set the autounseal data without being the leader. Ignoring the request."
+            )
+            return
+        if relation is None:
+            logger.warning("No relation found")
+            return
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return
+        credentials_secret_id = self._create_autounseal_credentials_secret(
+            relation, approle_id, approle_secret_id
+        )
+        relation.data[self.charm.app].update(
+            {
+                "address": vault_address,
+                "credentials_secret_id": credentials_secret_id,
+                "ca_certificate": ca_certificate,
+            }
+        )
+
+    def get_outstanding_requests(self, relation_id: Optional[int] = None) -> List[Relation]:
+        """Get the outstanding requests for the relation."""
+        outstanding_requests: List[Relation] = []
+        requirer_requests = self.get_requirer_requests(relation_id=relation_id)
+        for request in requirer_requests:
+            if not self._credentials_issued_for_request(relation_id=relation_id):
+                outstanding_requests.append(request)
+        return outstanding_requests
+
+    def get_requirer_requests(self, relation_id: Optional[int] = None) -> List[Relation]:
+        """Get all requests for the relation."""
+        pending: List[Relation] = []
+        relations = (
+            [
+                relation
+                for relation in self.model.relations[self.relation_name]
+                if relation.id == relation_id
+            ]
+            if relation_id is not None
+            else self.model.relations.get(self.relation_name, [])
+        )
+        for relation in relations:
+            assert isinstance(relation.app, ops.Application)
+            if not relation.active:
+                continue
+            app_data = relation.data[relation.app]
+            if not is_provider_data_valid(app_data):
+                continue
+            pending.append(relation)
+        return pending
+
+    def _credentials_issued_for_request(self, relation_id: Optional[int]) -> bool:
+        """Return whether credentials have been issued for the request."""
+        relation = self.model.get_relation(self.relation_name, relation_id)
+        if not relation:
+            return False
+        credentials = self.get_credentials(relation)
+        return all(credentials)
+
+    def get_credentials(self, relation) -> tuple[str | None, str | None]:
```

Comment:
```suggestion
    def get_credentials(self, relation: ops.Relation) -> tuple[str | None, str | None]:
```

Reply:
fixed in https://github.com/canonical/vault-k8s-operator/pull/346/commits/f46a5faef9cf31c600b380bf8e809225e9cceb3e